command/agent/config/config.go

```diff
@@ -350,6 +351,140 @@ func (c *Config) ValidateConfig() error {
+}
+
+func (c *Config) validateEnvTemplateConfig() error {
+	// if we are not in env-template mode, exit early
+	if c.Exec == nil && len(c.EnvTemplates) == 0 {
+		return nil
+	}
+
+	if c.Exec == nil {
+		return fmt.Errorf("'exec' element must be specified with 'env_template' entries")
+	}
+
+	if len(c.EnvTemplates) == 0 {
+		return fmt.Errorf("must specify at least one 'env_template' element with a top-level 'exec' element")
+	}
+
+	if c.ExitAfterAuth {
+		return fmt.Errorf("'exit_after_auth' cannot be specified with 'env_template' entries")
+	}
+
+	if c.APIProxy != nil {
+		return fmt.Errorf("'api_proxy' cannot be specified with 'env_template' entries")
+	}
+
+	if len(c.Templates) > 0 {
+		return fmt.Errorf("'template' cannot be specified with 'env_template' entries")
+	}
+
+	if c.Exec == nil || len(c.Exec.Command) == 0 {
+		return fmt.Errorf("'exec' requires non-empty 'command' field")
+	}
+
+	uniqueKeys := make(map[string]struct{})
+
+	for _, template := range c.EnvTemplates {
+		// only two fields are currently required (name & contents)
+		if template.MapToEnvironmentVariable == nil {
+			return fmt.Errorf("env_template: an environment variable name is required")
+		}
+
+		key := *template.MapToEnvironmentVariable
+
+		if _, exists := uniqueKeys[key]; exists {
+			return fmt.Errorf("env_template: duplicate environment variable name: %q", key)
+		}
+
+		uniqueKeys[key] = struct{}{}
+
+		if template.Contents == nil {
+			return fmt.Errorf("env_template[%s]: 'contents' is required", key)
+		}
+
+		if template.Backup != nil {
+			return fmt.Errorf("env_template[%s]: 'backup' is not allowed", key)
+		}
+
+		if template.Command != nil {
+			return fmt.Errorf("env_template[%s]: 'command' is not allowed", key)
+		}
+
+		if template.CommandTimeout != nil {
+			return fmt.Errorf("env_template[%s]: 'command_timeout' is not allowed", key)
+		}
+
+		if template.CreateDestDirs != nil {
+			return fmt.Errorf("env_template[%s]: 'create_dest_dirs' is not allowed", key)
+		}
+
+		if template.Destination != nil {
+			return fmt.Errorf("env_template[%s]: 'destination' is not allowed", key)
+		}
+
+		if template.ErrFatal != nil {
+			return fmt.Errorf("env_template[%s]: 'error_fatal' is not allowed", key)
+		}
+
+		if template.Exec != nil {
+			return fmt.Errorf("env_template[%s]: 'exec' is not allowed", key)
+		}
+
+		if template.Perms != nil {
+			return fmt.Errorf("env_template[%s]: 'perms' is not allowed", key)
+		}
+
+		if template.User != nil {
+			return fmt.Errorf("env_template[%s]: 'user' is not allowed", key)
+		}
+
+		if template.Uid != nil {
+			return fmt.Errorf("env_template[%s]: 'uid' is not allowed", key)
+		}
+
+		if template.Group != nil {
+			return fmt.Errorf("env_template[%s]: 'group' is not allowed", key)
+		}
+
+		if template.Gid != nil {
+			return fmt.Errorf("env_template[%s]: 'gid' is not allowed", key)
+		}
+
+		if template.Source != nil {
+			return fmt.Errorf("env_template[%s]: 'source' is not allowed", key)
```

Comment:
Why is this disallowed? It might be easier for customers to specify their environment variable templates as files and I see no reason to disallow this.

In general, I think we should aim to support as much as we can in the intersection between the two types of templating. I think it makes sense to only disallow config that doesn't apply for environment variable rendering.

Reply:
For the first iteration (beta release), we did not have the intention of supporting loading templates from files (though it may be easy to add with the current template implementation). We can definitely add it later if/when this is a requested feature. If it's super easy to implement, I could be convinced otherwise. CC @dhuckins 

> In general, I think we should aim to support as much as we can in the intersection between the two types of templating. I think it makes sense to only disallow config that doesn't apply for environment variable rendering.

I think we were trying to stick to the philosophy or releasing small and adding more later (if/when these features are needed). It is much easier to add functionality then to remove it once folks start using it.

Reply:
I think it'll be easy enough to add, but it's not something I feel so strongly about it'd be a blocker or anything, especially as we're near the release deadline. It might be as simple as just allowing it, as with the `left`/`right_delimiters`.

I think it looks a bit strange and perhaps inconsistent to support `contents` and `source` for one kind of template, but only `contents` for the other, without a technical limitation. If we're not doing this now, I think this is something we should look at supporting soon.

Reply:
Yea, that's a good point regarding the expectations! We'll try it out and see if it's easy to support or already supported

Reply:
Tested with our current build and it seems that `source` works without any changes so going to allow it as an alternative to `contents`.